File: packages/apple-sign-in/android/src/main/java/io/capawesome/capacitorjs/plugins/applesignin/classes/results/SignInResult.java
package io.capawesome.capacitorjs.plugins.applesignin.classes.results;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import io.capawesome.capacitorjs.plugins.applesignin.interfaces.Result;
import org.json.JSONObject;

public class SignInResult implements Result {

    @NonNull
    private final String authorizationCode;

    @NonNull
    private final String identityToken;

    @NonNull
    private final String user;

    @Nullable
    private final String email;

    @Nullable
    private final String givenName;

    @Nullable
    private final String familyName;

    @Nullable
    private final String state;

    public SignInResult(
        @NonNull String authorizationCode,
        @NonNull String identityToken,
        @NonNull String user,
        @Nullable String email,
        @Nullable String givenName,
        @Nullable String familyName,
        @Nullable String state
    ) {
        this.authorizationCode = authorizationCode;
        this.identityToken = identityToken;
        this.user = user;
        this.email = email;
        this.givenName = givenName;
        this.familyName = familyName;
        this.state = state;
    }

    @Override
    @NonNull
    public JSObject toJSObject() {
        JSObject result = new JSObject();
        result.put("authorizationCode", authorizationCode);
        result.put("identityToken", identityToken);
        result.put("user", user);
        result.put("email", email == null ? JSONObject.NULL : email);
        result.put("givenName", givenName == null ? JSONObject.NULL : givenName);
        result.put("familyName", familyName == null ? JSONObject.NULL : familyName);
        if (state != null) {
            result.put("state", state);
        }
        return result;
    }
}

File: packages/apple-sign-in/ios/Plugin/Classes/Results/SignInResult.swift
import Foundation
import Capacitor
import AuthenticationServices

@objc public class SignInResult: NSObject, Result {
    private let authorizationCode: String
    private let idToken: String
    private let user: String
    private let email: String?
    private let givenName: String?
    private let familyName: String?
    private let realUserStatus: String

    init(credential: ASAuthorizationAppleIDCredential) {
        self.authorizationCode = String(data: credential.authorizationCode ?? Data(), encoding: .utf8) ?? ""
        self.idToken = String(data: credential.identityToken ?? Data(), encoding: .utf8) ?? ""
        self.user = credential.user
        self.email = credential.email
        self.givenName = credential.fullName?.givenName
        self.familyName = credential.fullName?.familyName
        self.realUserStatus = SignInResult.mapRealUserStatus(credential.realUserStatus)
    }

    @objc public func toJSObject() -> AnyObject {
        var result = JSObject()
        result["authorizationCode"] = authorizationCode
        result["idToken"] = idToken
        result["user"] = user
        result["email"] = email == nil ? NSNull() : email
        result["givenName"] = givenName == nil ? NSNull() : givenName
        result["familyName"] = familyName == nil ? NSNull() : familyName
        result["realUserStatus"] = realUserStatus
        return result as AnyObject
    }

    private static func mapRealUserStatus(_ status: ASUserDetectionStatus) -> String {
        switch status {
        case .likelyReal:
            return "LIKELY_REAL"
        case .unknown:
            return "UNKNOWN"
        default:
            return "UNSUPPORTED"
        }
    }
}
